// bd2-ngx-hboxplot/esm2015/lib/hbox-plot/hbox-plot.dom.js
export class LookAndFeel {
    constructor() {
        this.vMargin = 25;
        this.hMarginL = 20;
        this.hMarginR = 15;
        this.rowWidth = 30;
        this.rowGap = 0.2;
        this.transitionTime = 600;
        this.boxStrokeWidth = '2px';
        this.boxFillOpacity = 0.35;
        this.meanStrokeWidth = '4px';
        this.labelFont = '12px';
        this.labelFillOpacity = 0.35;
        this.backLabelOpacity = 0.30;
        this.backdropColor = 'white';
        this.backdropOpacity = 1;
        this.whiskerStrokeWidth = '1px';
        this.outliersStrokeWidth = '1px';
        this.outliersCircleRadius = 3;
        this.outliersFillOpacity = 0.4;
    }
}
export let defualtLookAndFeel = function () {
    return new LookAndFeel();
};
export class GraphicContext {
    constructor() {
        /*transitionTime: number;
      
         get transitionOn(): boolean {
         return (this.transitionTime && this.transitionTime > 0);
         };*/
        this.labelsTimers = [];
    }
}
export function offsetScaleValue(x, pixOffset, scale) {
    const r = scale.range();
    const pos = scale(x) + pixOffset;
    if (pos < r[0]) {
        return r[0];
    }
    else if (pos > r[1]) {
        return r[1];
    }
    return pos;
}
